### PR TITLE
Add custom problem to dump action generated object syntax 

### LINKED ISSUE #8875: Dump input file syntax equivalent to action-generated objects
### Description of the enhancement or error report
Add ability to dump the equivalent input file syntax for all objects added by a given action. Note that this is _not_ what `--show-input` does.

### Rationale for the enhancement or information for reproducing the error
For debugging and input refactoring purposes.

### Identified impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #8875

## Example

```
./tensor_mechanics-opt -i tests/action/two_block_new.i Problem/type=DumpObjectsProblem Problem/dump_path=Modules/TensorMechanics/Master/block2
```

will output

```
[AuxKernels]
  [./strain_yy_block2]
    type = RankTwoAux
    block = 2
    execute_on = TIMESTEP_END
    index_i = 1
    index_j = 1
    rank_two_tensor = total_strain
    variable = strain_yy
  [../]
  [./stress_xx_block2]
    type = RankTwoAux
    block = 2
    execute_on = TIMESTEP_END
    index_i = 0
    index_j = 0
    rank_two_tensor = stress
    variable = stress_xx
  [../]
[]

[AuxVariables]
  [./strain_yy]
    blocks = '1 2'
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./stress_xx]
    blocks = '1 2'
    family = MONOMIAL
    order = CONSTANT
  [../]
[]

[Kernels]
  [./TM_block20]
    type = StressDivergenceTensors
    block = 2
    component = 0
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
    variable = disp_x
  [../]
  [./TM_block21]
    type = StressDivergenceTensors
    block = 2
    component = 1
    displacements = 'disp_x disp_y'
    use_displaced_mesh = true
    variable = disp_y
  [../]
[]

[Materials]
  [./block2_strain]
    type = ComputeFiniteStrain
    block = 2
    displacements = 'disp_x disp_y'
  [../]
[]

[Variables]
  [./disp_x]
    blocks = '1 2'
  [../]
  [./disp_y]
    blocks = '1 2'
  [../]
[]
```

([with a minimal set of input parameters](https://gist.github.com/dschwen/10c91443e1d88dcc5ed3fba050aa05f8/revisions)), which is what the

```
[Modules/TensorMechanics/Master]
  add_variables = true
  strain = FINITE
  generate_output = 'stress_xx'
  [./block2]
    block = 2
    additional_generate_output = 'strain_yy'
  [../]
[]
```

block in this input file creates.

### CODE REVIEW

**Comment by dschwen** [comment]
This is still a little rough around the edges. In particular I don't like that `bool` parameters are dumped as `0` or `1`, but that would need to be fixed in libMesh.

**Comment by dschwen** [comment]
Also detecting if a parameter is necessary, or could be omitted because the default value is identical to the set value is _surprisingly difficult_ in the InputParameters design. Ping @permcody

**Comment by dschwen** [comment]
I guess I could pull a fresh valid parameter set from the factory and compare the two (i.e. diff them)...

**Review by permcody — COMMENTED** [review]
I really like the output and honestly this is more useful than --show-input. However, I'm not sure I love the addition of a whole new problem just to get this done. I think I'd rather entertain adding in a callback in each of the `Add*` methods in the existing FEProblem class, or adding more information to the Syntax object to produce this output. That way you wouldn't have to switch systems to dump output and you could do it as part of a normal run. I'm all for replacing --show-input, it's pretty much useless as is.

**Comment by permcody** [comment]
> Also detecting if a parameter is necessary, or could be omitted because the default value is identical to the set value is surprisingly difficult in the InputParameters design.

This is a difficult problem since we support rich types and POD types. I don't believe the existing Parameters implementation requires that types have equality comparisons. If that's something we want to enforce, we'll we can go down that path.

**Comment by dschwen** [comment]
This does not yet capture objects added by actions added by the selected action. That would be important for simple initial conditions for example! I could build a tree of actions and their generated actions in the `ActionWarehouse::addActionBlock` method, and could aggregate the moose object syntax by the selected root action and all generated actions.

**Comment by YaqiWang** [comment]
How about moving this into `Action`? Users can call `addVariable` provided by Action instead of `_problem->addVariable`. We can then set a flag to turn the dump on/off easily and not need a new `FEProblem`. I am interested in this because we already had something similar in our derived Action, which is currently using to count the number of objects the Action added.

**Comment by dschwen** [comment]
Yaqi, I think that would not allow you to capture generated actions (or at least not let you bunch them up nicely in top level blocks). Also we would have to change _all_ actions ever made to support this new feature.

**Comment by YaqiWang** [comment]
Do you have a meta action? I am trying to understand what you are saying. Also you want to only dump inputs generated by a particular action? If so, I think this approach is doable. You can have a parameter in action `debug`, when it is true, we enable its dump capability and when a new action is created by this action, we pass `debug` to the created action. Did I miss something?

**Comment by dschwen** [comment]
Meta-actions are more common than you think! Just look at `AddVariableAction`. What I want is a refactoring tool for input files. 

* Explode Action into vanilla MooseObjects
* Remove unnecessary parameters

I want to run a moose executable on a given input, just like we run it to dump the syntax. This should not be used during a regular simulation (like you don't use `--syntax` or `--dump` in regular simulations either).

**Comment by YaqiWang** [comment]
In my terminology, `AddVariableAction` is not a meta action. It does not create another action. I know what you want to do. I have thought of it before although I gave up because the generated dump is not useful to me anyway (too big, we have thousands of MooseObjects).

**Comment by dschwen** [comment]
Yaqi, **it does**! Look again!

**Comment by YaqiWang** [comment]
Oh, you are right ;-) I was not thinking the IC part. But it does not matter anyway.

**Comment by dschwen** [comment]
What do you mean _ it does not matter anyway_?!

**Comment by YaqiWang** [comment]
Sorry, interrupted by something. I mean my approach should work. Want to have a discuss off-line?

**Comment by dschwen** [comment]
@permcody. Is this a dead end PR or is there any chance of getting this merge ready?

**Comment by permcody** [comment]
@dschwen - I like the output but I still think we should still incorporate this into `FEProblemBase`. I don't like the idea of having a separate problem that we have to maintain to keep this working.

**Comment by dschwen** [comment]
Well, the separate problem really only overrides the add calls. Not much
maintenance there.

**Comment by permcody** [comment]
Tag @idaholab/moose-team

**Comment by dschwen** [comment]
I understand that we don't want to add new problems that need to replicate certain initialization and execution loops. But this is not the case here. It is completely separate functionality. Adding that to FEProblemBase would just be clutter.

**Comment by dschwen** [comment]
So, for [tests/action/two_block_new.i](https://github.com/idaholab/moose/blob/devel/modules/tensor_mechanics/tests/action/two_block_new.i) this solution is capable of dumping the following paths:

```
AuxKernels/strain_theta
AuxKernels/stress_theta
AuxVariables/strain_theta
AuxVariables/stress_theta
BCs/bottom
BCs/left
BCs/right
BCs/top
Materials/_elastic_stress1
Materials/_elastic_stress2
Materials/elasticity_tensor
Modules/TensorMechanics/Master/block1
Modules/TensorMechanics/Master/block2
```

If dumping all other paths (such as `Mesh`, `MeshModifiers/block1`, etc.) is necessary I'll have to do a bit more research.

**Comment by permcody** [comment]
Another month has already ticked by and still no love from the MOOSE team. Ready to let this one die?

**Comment by rwcarlsen** [comment]
My new parser work will also affect something like this - the input file
rendering would probably build a a syntax tree that the new parser then
would render.

**Comment by dschwen** [comment]
What about actions? Those are the whole point of this pr.

**Comment by rwcarlsen** [comment]
I was just saying we should build a proper getpot parser tree and then have the getpot parser render the exploded/expanded input file syntax rather than directly building the input file syntax manually (via strings/streams).  But that is for the (near?) future when we have a nice getpot tree API to build with.

**Comment by permcody** [comment]
@friedmud  - this is all you

**Comment by sapitts** [comment]
@novasr Here's the code Daniel's working on that would let you build the explicit input file blocks that the tensor mechanics master action currently hides/builds, which might help with your work...

**Comment by permcody** [comment]
Hmm 30 day limit has been exceeded here. I've already closed this one once due to inactivity...

**Comment by lindsayad** [comment]
I think the debugging usefulness of this PR is very high; this new MOOSE team member is willing to show it some love :-)

**Comment by permcody** [comment]
"Hey Siri! Set a 30 day timer starting now"
*beep* *beep*
Confirmed, 30 day timer starting now...

File: framework/src/base/DumpObjectsProblem.C [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+#include <sstream>
+
+#include "libmesh/string_to_enum.h"
+
+template <>
+InputParameters
+validParams<DumpObjectsProblem>()
+{
+  InputParameters params = validParams<FEProblem>();
+  params.addRequiredParam<std::string>(
```
Comment (lindsayad) [inline framework/src/base/DumpObjectsProblem.C]: Could we do a vector of strings here to allow multiple action paths?

Comment (dschwen) [reply framework/src/base/DumpObjectsProblem.C]: Sure thing, I think all action expansions are build and stored anyways.

Comment (dschwen) [reply framework/src/base/DumpObjectsProblem.C]: And we could allow leaving the vector empty to output everything.

**Comment by aeslaughter** [comment]
I still feel this is the wrong approach, I would like to see a `MooseDump` Output object, in the output method do something I have below. 

This approach has two advantages, at least to me:
1. If a new object is added it is only a one line change to add it to the dump.
2. It works like other debug output objects, it can be spooled up in the Outputs block or we can 
   add a debug flag: `Debug/dump_objects=true`.

```
void output(...)
{
  dumpParams<Kernel>(_problem.getKernelWarehouse());
  ...
  dumpParams<OtherObjectType>(_problem.getOtherObjectTypeWarehouse());
};
```

**Comment by dschwen** [comment]
Not everything that needs to be dumped is in warehouses (mesh, variables). I'm having a deja vu here...

**Comment by aeslaughter** [comment]
Then handle the stuff that is not in a warehouse with functions.

```
dumpVariables(...)
dumpMesh(...)
dumpParam<Kernel>(...)
```

This PR also doesn't handle all the type of objects, what about Markers, Indicators, Dampers, Distributions, Samplers?

**Comment by YaqiWang** [comment]
I think it is better to let `ActionWarehouse` do this dumping, i.e. move `dumpHelper` stuff in `ActionWarehouse`.

**Comment by aeslaughter** [comment]
This is a debug output, it should be an Output object and there is no technical reason it can't be. I will even take this and do what is needed if you want.

**Comment by dschwen** [comment]
JSON output and YAML dumps are not "Outputs" either. I'll just unsubscribe from this and you do with it what you want.

**Comment by aeslaughter** [comment]
I am just saying what I think it should be, it is up to @permcody to make the decision. If he thinks that the Problem object is the correct approach, great, then this is done.

File: framework/src/base/DumpObjectsProblem.C [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+      if (!blocks.empty())
+        blocks += ' ';
+
+      blocks += subdomain_name;
+    }
+
+    if (active_subdomains->size() > 1)
+      blocks = "'" + blocks + "'";
+
+    param_text += "    blocks = " + blocks + '\n';
```
Comment (lindsayad) [inline framework/src/base/DumpObjectsProblem.C]: Ideally, it would be nice to just be able to plug output directly into input. `blocks` should be `block`

**Comment by andrsd** [comment]
Let me throw a monkey wrench into this PR ;-) In RELAP, we set a quite complicated equations (easily 3 different sets coupled via heat transfers), so lots of kernels, BCs, constraints. All done via components.  So, 1) if this only handles actions - it would be useless to us. If we got that info from warehouses, then we could benefit from this.  But, special kind of a problem is not an options for us, because we already have RELAP7Problem to handle the components and whatnot.  What do you guys think about a command line switch that would work together with `-i` and would dump the objects that were built by MOOSE?

**Comment by YaqiWang** [comment]
At least currently, adding an eigen kernel will result adding two kernels in reality. If we use object warehouse, we will print both. The point is that one single input syntax may produce multiple objects for purpose of capability...

**Comment by permcody** [comment]
Closing once again per the 30-day limit. I take no pleasure in this. I just believe that there is a better way. @aeslaughter proposes we do this by dumping from the warehouses directly instead of overriding the limited (and incomplete) number of "add" interfaces.

**Comment by novasr** [comment]
What's the path forward? Anyone still working on it?

**Comment by aeslaughter** [comment]
The path forward is provided in #10504, I am not working on it at the moment but would like to finish it up in the near future. Just need to find some time and money.

**Comment by novasr** [comment]
Thanks Andrew.

**Comment by dschwen** [comment]
📣 🔔 In honor of the **one year** anniversary 🎂🎈 of this PR, let me reopen 👻 it with a fresh 🍨 commit 💝 to bring it up to speed with recent MOOSE changes. 🍸 

@novasr, I suggest you check out this branch with:

```
git fetch upstream pull/8876/head:dumpproblem_8875
git checkout dumpproblem_8875
```

and build Bison with it to answer your specific question from the email.

**Comment by dschwen** [comment]
Oh, great, reopening doesn't work. :-|

**Comment by dschwen** [comment]
No worries guys! I opened a new PR!!! 😄

**Comment by dschwen** [comment]
🐱 😹 😸

File: framework/include/base/DumpObjectsProblem.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
```
Comment (permcody) [inline framework/include/base/DumpObjectsProblem.h]: Wrong header... What's wrong with you?

**Comment by permcody** [comment]
30 day timer starting now!

File: framework/include/base/DumpObjectsProblem.h (hunk 3 above)
Comment (dschwen) [reply framework/include/base/DumpObjectsProblem.h]: Retro style!

Comment (dschwen) [reply framework/include/base/DumpObjectsProblem.h]: @friedmud, @permcody is being mean to me in a review!

**Comment by permcody** [comment]
oh... must read carefully... new PR...

**Pull request closed without merging**